File: neet_code_all_cpp/sliding_window/max_num_of_vowels_k.cpp
#include <climits>
#include <iostream>
#include <string>
#include <unordered_set>

class Solution {
    public:
        // AC
        int maxVowels(std::string s, int k) {
            std::unordered_set<char> vowel_set = {'a', 'e', 'i', 'o', 'u'};
            int left = 0;
            int total = 0;
            int result = 0;

            for(int i = 0; i < k; ++i) {
                if(vowel_set.find(s[i]) != vowel_set.end()) {
                    ++total;
                }
                result = total;
            }

            for(int i = k; i < s.length(); ++i) {
                if(vowel_set.find(s[i]) != vowel_set.end()) {
                    ++total;
                }
                if(vowel_set.find(s[left]) != vowel_set.end()) {
                    --total;
                }

                ++left;
                result = std::max(result, total);
            }

            return result;
        }

        // AC
        int maxVowels2(std::string s, int k) {
            auto is_vowels = [](char& c) {
                if(c == 'a'
                        || c == 'e'
                        || c == 'i' 
                        || c == 'o'
                        || c == 'u' 
                  ){
                    return true;
                }

                return false;
            };

            int left = 0;
            int total = 0;
            int result = 0;

            for(int i = 0; i < k; ++i) {
                if(is_vowels(s[i])) {
                    ++total;
                }
                result = total;
            }

            for(int i = k; i < s.length(); ++i) {
                if(is_vowels(s[i])) {
                    ++total;
                }
                if(is_vowels(s[left])) {
                    --total;
                }

                ++left;
                result = std::max(result, total);
            }

            return result;
        }
};

// 模範解答
class SolutionAns {
    public:
        bool isVowel(char ch) {
            if(ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u') {
                return true;
            }

            return false;
        }

        int maxVowels(std::string s, int k) {
            int part_1 = 0, part_2 = 0;
            int count = 0, max_i = INT_MIN;

            while(part_2 < s.size()) {
                if(part_2 - part_1 == k) {

                    max_i = std::max(count, max_i);
                    if(max_i == k) {
                        return count;
                    }

                    if(isVowel(s[part_1++])) {
                        --count;
                    }
                    if(isVowel(s[part_2++])) {
                        ++count;
                    }

                } else {
                    if(isVowel(s[part_2])) {
                        ++count;
                    }

                    ++part_2;
                }
            }

            max_i = std::max(count, max_i);
            return max_i;
        }
};

int main(void) {
    using Case = std::pair<std::string, int>;
    Case case_1 = {"abciiidef", 3};
    // => 3
    Case case_2 = {"aeiou", 2};
    // => 2
    Case case_3 = {"leetcode", 3};
    // => 2
    Case case_4 = {"tryhard", 4};
    // => 1

    Solution s_1;

    std::cout << s_1.maxVowels(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.maxVowels(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.maxVowels(case_3.first, case_3.second) << std::endl;
    std::cout << s_1.maxVowels(case_4.first, case_4.second) << std::endl;

    std::cout << s_1.maxVowels2(case_1.first, case_1.second) << std::endl;
    std::cout << s_1.maxVowels2(case_2.first, case_2.second) << std::endl;
    std::cout << s_1.maxVowels2(case_3.first, case_3.second) << std::endl;
    std::cout << s_1.maxVowels2(case_4.first, case_4.second) << std::endl;

    SolutionAns s_ans;

    std::cout << s_ans.maxVowels(case_1.first, case_1.second) << std::endl;
    std::cout << s_ans.maxVowels(case_2.first, case_2.second) << std::endl;
    std::cout << s_ans.maxVowels(case_3.first, case_3.second) << std::endl;
    std::cout << s_ans.maxVowels(case_4.first, case_4.second) << std::endl;
}

File: neet_code_all/src/sliding_window/max_num_of_vowels_k.rs
use std::collections::HashSet;

// RustだとHashSetでなくて単に条件分岐で母音か否かの判定をしたほうが高速だった(C++はあまり変わらなかった)
struct Solution {}
impl Solution {
    // TLE
    // O(kN)
    pub fn max_vowels(s: String, k: i32) -> i32 {
        let s: Vec<char> = s.chars().collect();
        let k = k as usize;

        let vowel_set = HashSet::from(['a', 'e', 'i', 'o', 'u']);
        let mut result = 0;

        for i in 0..s.len() {
            let mut tmp_num = 0;

            let mut j = i;
            while j < s.len() && j < i + k {
                if vowel_set.contains(&s[j]) {
                    tmp_num += 1;
                }
                j += 1;
            }

            result = result.max(tmp_num);
        }

        result
    }

    // AC
    // O(N)
    pub fn max_vowels_2(s: String, k: i32) -> i32 {
        let s: Vec<char> = s.chars().collect();
        let k = k as usize;

        let vowel_set = HashSet::from(['a', 'e', 'i', 'o', 'u']);
        let (mut left, mut total) = (0, 0);
        let mut result = 0;

        for i in 0..k {
            if vowel_set.contains(&s[i]) {
                total += 1;
            }
            //println!("total(0): {}", total);
            result = total;
        }

        for j in k..s.len() {
            if vowel_set.contains(&s[j]) {
                total += 1;
            }
            //println!("total(1): {}", total);

            if vowel_set.contains(&s[left]) {
                total -= 1;
            }
            //println!("total(2): {}", total);

            left += 1;
            result = result.max(total);
        }

        result
    }

    // O(N): HashSetを使わない実装
    // AC
    pub fn max_vowels_3(s: String, k: i32) -> i32 {
        let s: Vec<char> = s.chars().collect();
        let k = k as usize;

        let is_vowels = |&c| match c {
            'a' | 'e' | 'i' | 'o' | 'u' => true,
            _ => false,
        };

        let (mut left, mut total) = (0, 0);
        let mut result = 0;

        for i in 0..k {
            if is_vowels(&s[i]) {
                total += 1;
            }
            //println!("total(0): {}", total);
            result = total;
        }

        for j in k..s.len() {
            if is_vowels(&s[j]) {
                total += 1;
            }
            //println!("total(1): {}", total);

            if is_vowels(&s[left]) {
                total -= 1;
            }
            //println!("total(2): {}", total);

            left += 1;
            result = result.max(total);
        }

        result
    }
}

// AC
// C++の模範解答より
struct SolutionAnsCpp {}
impl SolutionAnsCpp {
    pub fn max_vowels(s: String, k: i32) -> i32 {
        let is_vowel = |c| match c {
            'a' | 'e' | 'i' | 'o' | 'u' => true,

            _ => false,
        };

        let k = k as usize;
        let s: Vec<char> = s.chars().collect();

        let (mut part_1, mut part_2) = (0, 0);
        let (mut count, mut max_i) = (0, std::usize::MIN);

        while part_2 < s.len() {
            if part_2 - part_1 == k {
                max_i = max_i.max(count);

                if max_i == k {
                    return count as i32;
                }

                if is_vowel(s[part_1]) {
                    count -= 1;
                }
                part_1 += 1;

                if is_vowel(s[part_2]) {
                    count += 1;
                }
                part_2 += 1;
            } else {
                if is_vowel(s[part_2]) {
                    count += 1;
                }

                part_2 += 1;
            }
        }

        max_i = max_i.max(count);
        max_i as i32
    }
}

fn main() {
    let case_1 = ("abciiidef".to_string(), 3);
    // => 3
    let case_2 = ("aeiou".to_string(), 2);
    // => 2
    let case_3 = ("leetcode".to_string(), 3);
    // => 2
    let case_4 = ("tryhard".to_string(), 4);
    // => 1

    println!(
        "case_1: {}",
        Solution::max_vowels(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::max_vowels(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::max_vowels(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        Solution::max_vowels(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        Solution::max_vowels_2(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::max_vowels_2(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::max_vowels_2(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        Solution::max_vowels_2(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        Solution::max_vowels_3(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        Solution::max_vowels_3(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        Solution::max_vowels_3(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        Solution::max_vowels_3(case_4.0.clone(), case_4.1)
    );

    println!(
        "case_1: {}",
        SolutionAnsCpp::max_vowels(case_1.0.clone(), case_1.1)
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::max_vowels(case_2.0.clone(), case_2.1)
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::max_vowels(case_3.0.clone(), case_3.1)
    );
    println!(
        "case_4: {}",
        SolutionAnsCpp::max_vowels(case_4.0.clone(), case_4.1)
    );
}
